src/m365/spo/commands/group/group-member-add.ts

```diff
@@ -211,104 +202,55 @@ class SpoGroupMemberAddCommand extends SpoCommand {
 
     const validUserNames: string[] = [];
     const invalidUserNames: string[] = [];
-    const userIdentifiers: string = args.options.userName || args.options.email || args.options.aadGroupId! || args.options.aadGroupName! || args.options.userId!.toString();
+    const identifiers: string = args.options.userName || args.options.email || args.options.aadGroupId! || args.options.aadGroupName! || args.options.userId!.toString();
 
     return Promise
-      .all(userIdentifiers.split(',').map(async userIdentifier => {
-        const user = userIdentifier.trim();
+      .all(identifiers.split(',').map(async identifier => {
+        const trimmedIdentifier = identifier.trim();
         try {
           if (args.options.userId) {
-            await this.spoUserGet(args.options, user, logger, validUserNames);
+            if (this.verbose) {
+              logger.logToStderr(`Get UPN from SharePoint for user ${trimmedIdentifier}`);
+            }
+            const spoUser = await spo.getUserById(args.options.webUrl, trimmedIdentifier);
+            validUserNames.push(spoUser.UserPrincipalName);
+          }
+          else if (args.options.userName) {
+            validUserNames.push(trimmedIdentifier);
+          }
+          else if (args.options.aadGroupId) {
+            validUserNames.push(trimmedIdentifier);
           }
-          else if (args.options.aadGroupId || args.options.aadGroupName) {
-            await this.aadGroupGet(args.options, user, logger, validUserNames);
+          else if (args.options.aadGroupName) {
+            if (this.verbose) {
+              logger.logToStderr(`Get UPN from Azure AD for group ${trimmedIdentifier}`);
+            }
+            const group = await aadGroup.getGroupByDisplayName(trimmedIdentifier);
+            validUserNames.push(group.id!);
           }
           else {
-            await this.aadUserGet(args.options, user, logger, validUserNames);
+            if (this.verbose) {
+              logger.logToStderr(`Get UPN from Azure AD for user ${trimmedIdentifier}`);
+            }
+            const upn = await aadUser.getUserUpnByEmail(trimmedIdentifier);
+            validUserNames.push(upn);
           }
         }
         catch (err: any) {
-          logger.logToStderr(err.stderr);
-          invalidUserNames.push(userIdentifier);
+          invalidUserNames.push(identifier);
 
           return err;
         }
       }))
       .then((): Promise<string[]> => {
         if (invalidUserNames.length > 0) {
-          return Promise.reject(`Users not added to the group because the following users don't exist: ${invalidUserNames.join(', ')}`);
+          return Promise.reject(`Resource not added to the group because the following resources don't exist: ${invalidUserNames.join(', ')}`);
```

Comment:
I know you didn't write this code, but it's hurting my eyes quite badly. Could we please write this a bit cleaner?
I suggest that we keep the `Promise.all` to retrieve all identities, but throw an error immediately when retrieving one identity failed (Promise.all will instantly reject when one of its promises rejects).
Let's definitely get rid of this last `then` block.

Reply:
I hope your eyes feel better now ;-)